gui/widgets/groupedlist.cpp

```diff
@@ -0,0 +1,427 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "common/system.h"
+#include "common/frac.h"
+#include "common/tokenizer.h"
+
+#include "gui/widgets/groupedlist.h"
+#include "gui/widgets/scrollbar.h"
+#include "gui/dialog.h"
+#include "gui/gui-manager.h"
+
+#include "gui/ThemeEval.h"
+
+#define kGroupTag -2
+#define isGroupHeader(x) (x <= kGroupTag)
+#define indexToGroupID(x) (kGroupTag - x)
+
+namespace GUI {
+
+GroupedListWidget::GroupedListWidget(Dialog *boss, const String &name, const Common::U32String &tooltip, uint32 cmd)
+	: ListWidget(boss, name, tooltip, cmd) {
+
+}
+
+void GroupedListWidget::setList(const U32StringArray &list, const ColorList *colors) {
+	if (_editMode && _caretVisible)
+		drawCaret(true);
+
+	// Copy everything
+	_dataList = list;
+	_list = list;
+
+	_filter.clear();
+	_listIndex.clear();
+	_listColors.clear();
+
+	if (colors) {
+		_listColors = *colors;
+		assert(_listColors.size() == _dataList.size());
+	}
+
+	int size = list.size();
+	if (_currentPos >= size)
+		_currentPos = size - 1;
+	if (_currentPos < 0)
+		_currentPos = 0;
+	_selectedItem = -1;
+	_editMode = false;
+	g_system->setFeatureState(OSystem::kFeatureVirtualKeyboard, false);
+	groupByAttribute();
+	scrollBarRecalc();
+}
+
+void GroupedListWidget::setAttributeValues(const U32StringArray &attrValues) {
+	_attributeValues = attrValues;
+	if (!attrValues.empty())
+		assert(_attributeValues.size() == _dataList.size());
+}
+
+void GroupedListWidget::append(const String &s, ThemeEngine::FontColor color) {
+	if (_dataList.size() == _listColors.size()) {
+		// If the color list has the size of the data list, we append the color.
+		_listColors.push_back(color);
+	} else if (!_listColors.size() && color != ThemeEngine::kFontColorNormal) {
+		// If it's the first entry to use a non default color, we will fill
+		// up all other entries of the color list with the default color and
+		// add the requested color for the new entry.
+		for (uint i = 0; i < _dataList.size(); ++i)
+			_listColors.push_back(ThemeEngine::kFontColorNormal);
+		_listColors.push_back(color);
+	}
+
+	_dataList.push_back(s);
+	_list.push_back(s);
+
+	setFilter(_filter, false);
+
+	scrollBarRecalc();
+}
+
+void GroupedListWidget::setGroupHeaderFormat(const U32String &prefix, const U32String &suffix) {
+	_groupHeaderPrefix = prefix;
+	_groupHeaderSuffix = suffix;
+}
+
+void GroupedListWidget::groupByAttribute() {
+	_groupExpanded.clear();
+	_groupHeaders.clear();
+	_groupValueIndex.clear();
+	_itemsInGroup.clear();
+
+	if (_attributeValues.empty()) {
+		_groupExpanded.push_back(true);
+		_groupHeaders.push_back(String("All"));
+		_groupValueIndex.setVal(String("All"), 0);
+		for (uint i = 0; i < _dataList.size(); ++i) {
+			_itemsInGroup[0].push_back(i);
+		}
+		sortGroups();
+		return;
+	}
+
+	for (uint i = 0; i < _dataList.size(); ++i) {
+		int groupID;
```

Comment:
Uninitialized here, and not referenced before line 132, move the declaration down to line 132 to make it obvious that it's initialized.

Reply:
Fixed in 324d128 by declaring and initializing on the same line.